### ISSUE TITLE
JWT: Works with Safari?

### ISSUE THREAD

**Issue by netwire88** [Shopify/shopify_app#1101]
Hi,
I've had some success implementing JWT on Chrome per https://github.com/Shopify/turbolinks-jwt-sample-app. However, in Safari Version 13.0.1, I'm getting error `Blocked autofocusing on a form control in a cross-origin subframe.` and it's redirecting back to Shopify login.
Does v 15.0 of shopify_app work with JWT + Safari?

**Comment by netwire88** [comment]
Any update?

**Comment by tanema** [comment]
Hello! It seems like safari doesn't forward custom headers when it follows redirects. We have some pending doc updates to help support this, but in this case, this feature is still somewhat in-development

**Comment by netwire88** [comment]
Thanks, so as of now, JWT doesn't work on Safari?

**Comment by netwire88** [comment]
Any update on this?

**Comment by netwire88** [comment]
@tanema, I just tried using v16.1 in Safari and am still getting error `Unrecognized Content-Security-Policy directive 'worker-src'.`
If it helps, it then raises this error:

<img width="1024" alt="2021-01-08_14-43-43" src="https://user-images.githubusercontent.com/1209073/104057315-ef47de00-51bf-11eb-80a6-827d1bd58d2b.png">

If it helps, here's the stacktrace:

```
shopify_app (16.1.0) lib/shopify_app/controller_concerns/login_protection.rb:198:in `current_shopify_domain' 
shopify_app (16.1.0) app/controllers/concerns/shopify_app/ensure_authenticated_links.rb:14:in `redirect_to_splash_page' 
activesupport (6.0.3.4) lib/active_support/callbacks.rb:428:in `block in make_lambda' 
activesupport (6.0.3.4) lib/active_support/callbacks.rb:181:in `block (2 levels) in halting_and_conditional' 
actionpack (6.0.3.4) lib/abstract_controller/callbacks.rb:34:in `block (2 levels) in <module:Callbacks>' 
activesupport (6.0.3.4) lib/active_support/callbacks.rb:182:in `block in halting_and_conditional' 
activesupport (6.0.3.4) lib/active_support/callbacks.rb:513:in `block in invoke_before' 
activesupport (6.0.3.4) lib/active_support/callbacks.rb:513:in `each' 
activesupport (6.0.3.4) lib/active_support/callbacks.rb:513:in `invoke_before' 
activesupport (6.0.3.4) lib/active_support/callbacks.rb:110:in `block in run_callbacks' 
activesupport (6.0.3.4) lib/active_support/callbacks.rb:139:in `run_callbacks' 
actionpack (6.0.3.4) lib/abstract_controller/callbacks.rb:41:in `process_action' 
actionpack (6.0.3.4) lib/action_controller/metal/rescue.rb:22:in `process_action' 
actionpack (6.0.3.4) lib/action_controller/metal/instrumentation.rb:33:in `block in process_action' 
activesupport (6.0.3.4) lib/active_support/notifications.rb:180:in `block in instrument' 
activesupport (6.0.3.4) lib/active_support/notifications/instrumenter.rb:24:in `instrument' 
activesupport (6.0.3.4) lib/active_support/notifications.rb:180:in `instrument' 
actionpack (6.0.3.4) lib/action_controller/metal/instrumentation.rb:32:in `process_action' 
actionpack (6.0.3.4) lib/action_controller/metal/params_wrapper.rb:245:in `process_action' 
activerecord (6.0.3.4) lib/active_record/railties/controller_runtime.rb:27:in `process_action' 
```

**Comment by petebof** [comment]
Just checking the state of things.

- JWT does not work on Safari, right?

- The project Readme states: "**[Not recommended for embedded apps]** with-cookie-authentication - sets up the authentication strategy of the app to use cookies. By default, it uses JWT based session tokens."

If this is the case, is the recommendation to use a solution that does not work on Safari? Is cookie authentication still working on embedded apps?

**Comment by NabeelAhsen** [comment]
**TL;DR**  Following the release of the `EnsureAuthenticatedLinks` concern in `v16.1.0`, JWTs should be supported in Safari. There are no known issues with this solution yet, apart from known ITP-related cookies issues.

Hi @netwire88 couple questions:
- Are you attempting to visit your app in an embedded state? (i.e, does your URL match `some_domain.myshopify.com/admin/apps/your_app`
- Are you able to verify using a breakpoint like `byebug` or `pry` to see why your `current_shopify_domain` is failing?

I am able to recreate this exact issue and I have a couple of observations:

1. `Unrecognized Content-Security-Policy directive 'worker-src'` is likely unrelated to the issue. I see this in Safari while reproducing this issue and when it's working fine.
2. I've been able to reproduce this issue when i'm visiting my app in a non-embedded state (i.e outside of Shopify Admin). If this is similar to how you're visiting your app, then this is likely the issue.
3. Apps using session tokens should never be visited in a non-embedded state.

### My steps to reproduce this issue in Chrome + Safari
1. Visit a deep-linked page in a non-embedded state: `https://<ngrok_url>.com/deep_linked_resource`
2. The `AuthenticatedController` determines that there is no active session right now, and delegates this enforcement to the `EnsureAuthenticatedLinks` concern.
3. The `EnsureAuthenticatedLinks` attempts to get a valid session token by visiting your splash page controller. (note that this concern is only needed if your app uses Turbolinks)
4. The following line attempts to call `current_shopify_domain` on `LoginProtection` concern. At this stage, your `current_shopify_domain` should have been passed in automatically by Shopify Admin, but only if your app is in an embedded state.

https://github.com/Shopify/shopify_app/blob/6d2bb714a611b15d7366d3f60d53bb428dc23631/app/controllers/concerns/shopify_app/ensure_authenticated_links.rb#L14

https://github.com/Shopify/shopify_app/blob/6d2bb714a611b15d7366d3f60d53bb428dc23631/lib/shopify_app/controller_concerns/login_protection.rb#L179-L187

https://github.com/Shopify/shopify_app/blob/6d2bb714a611b15d7366d3f60d53bb428dc23631/lib/shopify_app/controller_concerns/login_protection.rb#L189-L191

The following fails:

https://github.com/Shopify/shopify_app/blob/6d2bb714a611b15d7366d3f60d53bb428dc23631/lib/shopify_app/controller_concerns/login_protection.rb#L204-L207

The resulting error (Note that the red highlighted box is unrelated to the issue you're facing, but it' worth pointing out that it's only shown on Safari):

<img width="1270" alt="Screen Shot 2021-01-12 at 10 44 26 AM" src="https://user-images.githubusercontent.com/13038454/104337722-c8d9b980-54c3-11eb-9a0c-2db46d58504f.png">

**Comment by netwire88** [comment]
Thanks @NabeelAhsen, this is helpful. Good to know the worker-src isn't an issue.
I'm loading the app in embedded state. However, as I dug a little deeper, it seems that this is failing in Safari after any `redirect_to any_name_path`.
After following a `redirect_to`, shopify_app triggers `EnsureAuthenticatedLinks` concern. However, `jwt_shopify_domain` seems to be blank. The oddity is that this isn't happening in Chrome but only in Safari.
I tried to test this out in https://github.com/Shopify/turbolinks-jwt-sample-app but I'm unable to get the app to work. Can you try a `redirect_to` within a Controller and see if it works?

**Comment by NabeelAhsen** [comment]
Hi @netwire88, as pointed out above Safari does not seem to be forwarding custom headers when following redirects. Specifically in our case, it does not forward the `X-Authorization` header on redirects. 
- `request.env['jwt.shopify_domain']` comes directly from the session token provided in this header, and so it explains why the `jwt_shopify_domain` is unavailable following a redirect in Safari.

--- 
The reason `EnsureAuthenticatedLinks` works the first time when you visit a deep-link, like `/your_app/products`, is because it redirects to the `SplashPageController`, which then tells Turbolinks to visit the `return_to` parameter rather than redirect to it: https://github.com/Shopify/turbolinks-jwt-sample-app/blob/78d5f6c2e50f4b9303c6046dcdea834bf2d1c70d/app/javascript/shopify_app/shopify_app.js#L18-L25
- This ensures that the `X-Authorization` header isn't lost, because it's not a redirect.
- This behaviour works across Chrome, Safari, Firefox, etc. 
- This behaviour does not work for subsequent redirects made within the app in Safari yet - only the initial redirect to `SplashPageController` that bootstraps `shopify_app.js`.


> I tried to test this out in https://github.com/Shopify/turbolinks-jwt-sample-app but I'm unable to get the example app to work. Can you try a redirect_to within any controller and see if it works in Safari?

Are there any steps in the [Quick Start](https://github.com/Shopify/turbolinks-jwt-sample-app#quick-start) section of that sample app that I can help clarify? 
- FYI: We've updated one section of this that suggests the wrong environment variable name.

**Comment by netwire88** [comment]
Thanks @NabeelAhsen , if Safari doesn't work with `request.env['jwt.shopify_domain']`, have you been able to get ability to Approve subscription via Shopify working in Safari? That requires redirect back using deep linking. I'm able to get it to work in Chrome but not Safari. Any alternatives?

**Comment by kirillplatonov** [comment]
I solved the redirect issue by replacing the redirect status code with 300 (ignored by browsers) and handling redirect in [Turbo](https://turbo.hotwire.dev/).
```ruby
# app/controllers/example_controller.rb

def create
  # ...

  if @campaign.save
    redirect_to campaigns_path, status: 300
  else
    # ...
  end
end
```
```javascript
// app/javascript/shopify_app/shopify_app.js

document.addEventListener("turbo:before-fetch-request", (event) => {
  const headers = event.detail.fetchOptions.headers;
  headers["Authorization"] = `Bearer ${window.sessionToken}`;
});

document.addEventListener("turbo:before-fetch-response", (event) => {
  const response = event.detail.fetchResponse;
  const status = response.statusCode;
  const location = response.header("Location");

  if (status === 300 && location !== null) {
    event.preventDefault();
    Turbo.visit(location);
  }
});
```
The same can be done with Turbolinks as well using `turbolinks:request-start`.

**Comment by denoyaro** [comment]
@flavio-b did you ever figure out in your solution why location was being visited twice?

**Comment by flavio-b** [comment]
@denoyaro Haven't figured it out, so I abandoned this solution. Safari users don't see flash messages, for now.

The event `turbolinks:request-end` fires AFTER the network request completes. Perhaps Turbolinks or the browser are responding to that request before it goes to the handler, but that's just my guess. The status 300 is rarely used, so I couldn't find much info about how the browser would respond to it.

**Comment by denoyaro** [comment]
@flavio-b thanks for the update. I'll keep looking out for other solutions. Hopefully I can find one and share here.
